### PR TITLE
Introduce operator origin stack trace in the error message

### PR DESCRIPTION
**Pull request by klecki**
## Description:
Adjust how executor formats context of exception message.
Add the formatted origin stack trace to the error message as a context.

Remove the C++ backtrace from being propagated as a string to the user.
Handle translating the DALIException thrown from DALI_ENFORCE to DaliError
hierarchy.
Add a baseline conversion to runtime_error for the top level DaliError (keeping in line
with the current defaults).

Use new facilities to present math ops as belonging to `math` api with a short display name,
Limit the stack trace that is reported there to not show the internal implementation.

Add tests for conditional split and merge and math ops.

### Affected modules and functionalities:
DALI_ENFORCE, error reporting.

### Tests:

- [x] Existing tests apply
All tests should pass, if there are tests that expect the error message in some specific form
they may need to be adjusted.
tf record reader test adjusted to check for the name
- [x] New tests added
  - [x] Python tests
Added tests for potentially problematic cases: arithm ops, conditionals 
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

File: dali/pipeline/operator/op_schema.h [hunk 1]
```diff
@@ -82,10 +82,20 @@ class DLL_PUBLIC OpSchema {
...
 
+  /**
+   * @brief Returns the module path of this operator.
+   */
+  DLL_PUBLIC const std::vector<std::string> &ModulePath() const;
+
+  /**
+   * @brief Returns the camel case name of the operator (without the module path)
+   */
+  DLL_PUBLIC const std::string &OperatorName() const;
```
Comment (mzient) [inline dali/pipeline/operator/op_schema.h:97]: How is that different from the schema name? Why do we need both?

File: dali/pipeline/operator/op_schema.h [hunk 2]
```diff
@@ -82,10 +82,20 @@ class DLL_PUBLIC OpSchema {
...
    */
   DLL_PUBLIC const std::string &name() const;
 
+  /**
+   * @brief Returns the module path of this operator.
+   */
+  DLL_PUBLIC const std::vector<std::string> &ModulePath() const;
+
+  /**
+   * @brief Returns the camel case name of the operator (without the module path)
```
Comment (mzient) [inline dali/pipeline/operator/op_schema.h:95]:
```suggestion
   * @brief Returns the PascalCase name of the operator (without the module path)
```

File: dali/pipeline/operator/op_schema.h (hunk 1 above)
Comment (klecki) [reply dali/pipeline/operator/op_schema.h:97]: Returns the camel case name of the operator *(without the module path)*

File: dali/pipeline/operator/name_utils.cc [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+std::string GetOpModule(const OpSpec &spec, ModuleSpecKind kind) {
+  DALI_ENFORCE(kind != ModuleSpecKind::OpOnly, "This function can be only queried about module");
+  std::string result;
+  if (kind == ModuleSpecKind::LibApiModule) {
+    result += "nvidia.dali.";
+    result += GetOpApi(spec);
+  } else if (kind == ModuleSpecKind::ApiModule) {
+    result += GetOpApi(spec);
+  }
```
[the remark this answers is not in the document]
Comment (klecki) [reply dali/pipeline/operator/name_utils.cc:37]: Done in the previous PR

File: dali/pipeline/operator/op_spec.h [hunk 4]
```diff
@@ -56,35 +56,34 @@ class DLL_PUBLIC OpSpec {
...
   /**
-   * @brief Sets the name of the Operator.
+   * @brief Sets the schema name of the Operator.
    */
-  DLL_PUBLIC inline void set_name(const string &name) {
-    name_ = name;
-    schema_ = name_.empty() ? nullptr : SchemaRegistry::TryGetSchema(name_);
+  DLL_PUBLIC inline void SetName(const string &schema_name) {
+    schema_name_ = schema_name;
+    schema_ = schema_name_.empty() ? nullptr : SchemaRegistry::TryGetSchema(schema_name_);
```
[the remark this answers is not in the document]
Comment (klecki) [reply dali/pipeline/operator/op_spec.h:82]: Adjusted, I'm not sure if we need the pointer variant.

**Comment by klecki** [comment]
!build

File: dali/pipeline/operator/op_spec.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (mzient) [reply dali/pipeline/operator/op_spec.h]: It's useful for tests. You can create an op-spec with an unregistered schema.

File: include/dali/core/error_handling.h [hunk 5]
```diff
@@ -234,28 +248,28 @@ inline dali::string GetStacktrace() {
 #define FILE_AND_LINE __FILE__ ":" DALI_STR(__LINE__)
 
 #define DALI_MESSAGE_WITH_STACKTRACE(str)\
-  (std::string("[" FILE_AND_LINE "] ") + str + dali::GetStacktrace())
+  (std::string("[" FILE_AND_LINE "] ")), (std::string(str)), dali::GetStacktrace()
```
Comment (mzient) [inline include/dali/core/error_handling.h:251]:
```suggestion
#define DALI_MESSAGE_WITH_STACKTRACE(...)\
  ("[" FILE_AND_LINE "] "), ##__VA_ARGS __, dali::GetStacktrace()
```
This way you allow the message to be empty, to be multi-part and of heterogeneous types.
Also, you save the unnecessary construction of an `std::string` out of a string literal.

File: include/dali/core/error_handling.h [hunk 6]
```diff
@@ -234,28 +248,28 @@ inline dali::string GetStacktrace() {
...
 
 #define DALI_MESSAGE(str)\
-  (std::string("[" FILE_AND_LINE "] ") + str)
+  (std::string("[" FILE_AND_LINE "] ")), (std::string(str))
```
Comment (mzient) [inline include/dali/core/error_handling.h:254]:
```suggestion
#define DALI_MESSAGE(...)\
  ("[" FILE_AND_LINE "] "), ##__VA_ARGS__
```

File: include/dali/core/error_handling.h [hunk 7]
```diff
@@ -234,28 +248,28 @@ inline dali::string GetStacktrace() {
...
+  do {                                                                              \
+    std::cerr << dali::make_string(DALI_MESSAGE_WITH_STACKTRACE(str)) << std::endl; \
   } while (0)
 
-#define DALI_WARN(...)                                                      \
-  do {                                                                      \
-    std::cerr << DALI_MESSAGE(dali::make_string(__VA_ARGS__)) << std::endl; \
+#define DALI_WARN(str)                                              \
+  do {                                                              \
+    std::cerr << dali::make_string(DALI_MESSAGE(str)) << std::endl; \
```
Comment (mzient) [inline include/dali/core/error_handling.h:266]: We should probably have some synchronized print... Stringification before printing helps a big, but not much.
```suggestion
    diag_print(DALI_MESSAGE(str));
```
where `diag_print` has a mutex.

File: include/dali/core/error_handling.h [hunk 8]
```diff
@@ -234,28 +248,28 @@ inline dali::string GetStacktrace() {
...
   } while (0)
 
-#define DALI_WARN_ONCE(...)                                                          \
-  do {                                                                               \
-    static int dummy =                                                               \
-        (std::cerr << DALI_MESSAGE(dali::make_string(__VA_ARGS__)) << std::endl, 0); \
+#define DALI_WARN_ONCE(str)                                                  \
+  do {                                                                       \
+    static int dummy =                                                       \
+        (std::cerr << dali::make_string(DALI_MESSAGE(str)) << std::endl, 0); \
```
Comment (mzient) [inline include/dali/core/error_handling.h:272]: Now that's clever :) Still, we don't need to produce a string to print a message.
```suggestion
    static int dummy =                                                       \
        (print(std::cerr, DALI_MESSAGE(str), std::endl), 0); \
```

File: dali/util/nvml_wrap.cc [hunk 9]
```diff
@@ -66,7 +66,7 @@ nvmlReturn_t nvmlInitChecked() {
 
   nvmlReturn_t ret = nvmlInit();
   if (ret != NVML_SUCCESS) {
-    DALI_WARN("nvmlInitChecked failed: ", nvmlErrorString(ret));
+    DALI_WARN(dali::make_string("nvmlInitChecked failed: ", nvmlErrorString(ret)));
```
Comment (mzient) [inline dali/util/nvml_wrap.cc:69]:
```suggestion
    DALI_WARN("nvmlInitChecked failed: ", nvmlErrorString(ret));
```
See my comments below.

File: dali/util/mmaped_file.cc [hunk 10]
```diff
@@ -102,7 +102,7 @@ static void *file_map(const char *path, size_t *length, bool read_ahead) {
 
 fail:
   if (p == nullptr) {
-    DALI_FAIL("File mapping failed: " + path);
+    DALI_FAIL(dali::make_string("File mapping failed: ", path));
```
Comment (mzient) [inline dali/util/mmaped_file.cc:105]:
```suggestion
    DALI_FAIL("File mapping failed: ", path);
```
See below.

File: include/dali/core/error_handling.h (hunk 5 above)
Comment (klecki) [reply include/dali/core/error_handling.h]: done

File: include/dali/core/error_handling.h (hunk 7 above)
Comment (klecki) [reply include/dali/core/error_handling.h]: I will consider this out of scope for now.

File: dali/util/nvml_wrap.cc (hunk 9 above)
Comment (klecki) [reply dali/util/nvml_wrap.cc]: Reworked to accept VA_ARGS

**Comment by klecki** [comment]
!build

**Comment by klecki** [comment]
!build

**Comment by klecki** [comment]
!build

**Comment by klecki** [comment]
!build

**Comment by klecki** [comment]
!build

File: dali/pipeline/executor/executor.cc [hunk 11]
```diff
@@ -190,7 +206,7 @@ void Executor<WorkspacePolicy, QueuePolicy>::RunCPUImpl(size_t iteration_id) {
       RunHelper(op_node, ws, iteration_id);
       FillStats(cpu_memory_stats_, ws, "CPU_" + op_node.instance_name, cpu_memory_stats_mutex_);
     } catch (...) {
-      HandleError("CPU", op_node);
+      HandleError(op_node);
```
Comment (mzient) [inline dali/pipeline/executor/executor.cc:209]: :+1:

File: dali/pipeline/operator/error_reporting.cc [hunk 12]
```diff
@@ -90,5 +94,22 @@ void PropagateError(ErrorInfo error) {
   }
 }
 
+std::string GetErrorContextMessage(const OpSpec &spec) {
+  auto stage = spec.GetArgument<std::string>("device");
```
Comment (mzient) [inline dali/pipeline/operator/error_reporting.cc:98]: Nit
```suggestion
  auto device = spec.GetArgument<std::string>("device");
```

File: dali/util/fits.cc [hunk 13]
```diff
@@ -287,10 +287,12 @@ void HandleFitsError(int status) {
...
-    DALI_WARN_ONCE("Loaded instance of CFITSIO library does not support multithreading. "
-                  "Please recompile CFITSIO in reentrant mode (--enable-reentrant) "
-                  "or use CFITSIO delivered in DALI_deps. Using non-reentrant version "
-                  "of CFITSIO may degrade the performance.");
+    static const char msg[] =
+        "Loaded instance of CFITSIO library does not support multithreading. "
+        "Please recompile CFITSIO in reentrant mode (--enable-reentrant) "
+        "or use CFITSIO delivered in DALI_deps. Using non-reentrant version "
+        "of CFITSIO may degrade the performance.";
+    DALI_WARN_ONCE(msg);
```
Comment (mzient) [inline dali/util/fits.cc:295]: Why?

File: dali/util/nvml_wrap.cc [hunk 14]
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) 2020-2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+// Copyright (c) 2020-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (mzient) [inline dali/util/nvml_wrap.cc:1]: Please revert?

File: dali/util/std_cufile.cc [hunk 15]
```diff
@@ -70,8 +71,8 @@ static void cufile_open(cufile::CUFileHandle& fh, size_t& length, const char* pa
 
   CUfileError_t status = cuFileHandleRegister(&(fh.cufh), &descr);
   if (status.err != CU_FILE_SUCCESS) {
-    DALI_FAIL("CUFile import failed: " + path + ". " +
-              std::string(cufileop_status_error(status.err)) + ".");
+    DALI_FAIL("CUFile import failed: ", path, ". ",
+              std::string(cufileop_status_error(status.err)), ".");
```
Comment (mzient) [inline dali/util/std_cufile.cc:75]:
```suggestion
              cufileop_status_error(status.err), ".");
```

**Review by mzient — APPROVED** [review]
OK, with some nitpicks.

File: include/dali/core/error_handling.h [hunk 16]
```diff
@@ -233,29 +247,32 @@ inline dali::string GetStacktrace() {
 #define DALI_STR(x) DALI_STR2(x)
 #define FILE_AND_LINE __FILE__ ":" DALI_STR(__LINE__)
 
-#define DALI_MESSAGE_WITH_STACKTRACE(str)\
-  (std::string("[" FILE_AND_LINE "] ") + str + dali::GetStacktrace())
+#define DALI_MESSAGE_WITH_STACKTRACE(...)\
+  (std::string("[" FILE_AND_LINE "] ")), ##__VA_ARGS__, dali::GetStacktrace()
```
Comment (mzient) [inline include/dali/core/error_handling.h:251]:
```suggestion
#define DALI_MESSAGE_AND_STACKTRACE(...)\
  (std::string("[" FILE_AND_LINE "] ")), ##__VA_ARGS__, dali::GetStacktrace()
```
Now it's no longer a message with stack trace - those are two separate things.

Comment (klecki) [reply include/dali/core/error_handling.h:251]: done

File: dali/util/std_cufile.cc (hunk 15 above)
Comment (klecki) [reply dali/util/std_cufile.cc:75]: done

File: dali/util/fits.cc (hunk 13 above)
Comment (klecki) [reply dali/util/fits.cc:295]: Due to intermediate change in the macros implementation.

File: dali/pipeline/operator/error_reporting.cc (hunk 12 above)
Comment (klecki) [reply dali/pipeline/operator/error_reporting.cc:98]: done

**Comment by klecki** [comment]
!build

**Review by awolant — APPROVED** [review]

**Comment by klecki** [comment]
!build

**Comment by klecki** [comment]
!build

**Pull request merged**